### PR TITLE
test: zoe null upgrade by swingset.CoreEval governance decision

### LINKED ISSUE #7771: test zoe null upgrade during running vaults system
## What is the Problem Being Solved?

Upgrading zoe in isolation is reasonably well tested, but I'm not aware that we've tested such an upgrade while the whole system is running. Do all the relevant parts survive the upgrade trauma?

## Test Plan / Design

 - [x] In the `docker-upgrade-test` context, make a core eval proposal to do a null upgrade of zoe.
 - [x] check that opening a vault still works
 - [ ] review logs and open issues for things such as seats that should have been garbage-collected but weren't exited (more: https://github.com/Agoric/agoric-sdk/issues/7771#issuecomment-1562009781)
 - [x] integrate this test into ci

## Security Considerations

Knowing that this works helps us understand our options should various risks arise.

## Scaling Considerations

none?

### PR DESCRIPTION
**Pull request by dckc**
closes: #7771

## Description

Using the docker upgrade-tests framework, test that we can upgrade zoe using swingset.CoreEval.

 - pre test: kernel DB shows zoe vat is at incarnation 0
 - action: use CoreEval to upgrade zoe
 - test: kernel DB shows zoe vat is at incarnation 1

### Security Considerations

simulates a BLD staker vote

### Scaling Considerations

none?

### Documentation Considerations

provides another example of using CoreEval (cc @Tyrosine22 )

### Testing Considerations

- [x] integrate with upgrade-tests

### CODE REVIEW

File: packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/zoe-upgrade/vat-status.mjs [hunk 1]
```diff
@@ -0,0 +1,105 @@
+// @ts-check
+import processAmbient from 'process';
+import dbOpenAmbient from 'better-sqlite3';
+
+/**
+ * @file look up vat incarnation from kernel DB
```
Comment (dckc) [inline packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/zoe-upgrade/vat-status.mjs:6]: @warner does it seem like a good idea for a test to do this "clean room" style implementation of querying the kernel DB?

cc @raphdev

File: packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/zoe-upgrade/zoe-upgrade-driver.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+here='upgrade-test-scripts/agoric-upgrade-11/zoe-upgrade'
+
+agd --chain-id=agoriclocal \
+  tx gov submit-proposal swingset-core-eval \
+  ${here}/zoe-upgrade-permit.json ${here}/zoe-upgrade-script.js \
+  --title="Zoe Upgrade" --description="zoe upgrade test" \
+  --deposit=10000000ubld \
+  --gas=auto --gas-adjustment=1.2 \
+  --yes -o json --from=validator --keyring-backend=test -b sync
```
Comment (dckc) [inline packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/zoe-upgrade/zoe-upgrade-driver.sh]: Looking at other `tx gov submit` stuff in this framework, I think a few of these flags are over-kill in this context. They were needed for devnet or something.

**Comment by dckc** [comment]
@michaelfig are we ready to close #7771 when this lands? Or are there some issues we should raise based on what we saw in [zoe-upgrade-log.txt](https://gist.github.com/dckc/0f702f12037a8a19c2bca9e85842b2c3#file-zoe-upgrade-log-txt)? I took another look and I can't quite tell why smartWallet stuff is happening.

**Comment by michaelfig** [comment]
> @michaelfig are we ready to close #7771 when this lands?

Yes, I think this PR closes that issue.

> Or are there some issues we should raise based on what we saw in [zoe-upgrade-log.txt](https://gist.github.com/dckc/0f702f12037a8a19c2bca9e85842b2c3#file-zoe-upgrade-log-txt)?

We should create issues around:

- need to characterise why (and fix) long-lived offers held across zoe upgrades are exited instead of the upgrade being unobservable to external parties
- offer handlers that drop the seat without throwing or rejecting (i.e. incorrect contracts) should be fixed to explicitly `seat.exit()`.  This will mask the above problem, so we should do it only after we've managed to fix that one.

> I took another look and I can't quite tell why smartWallet stuff is happening.

I think the only other mysterious smartWallet log happens because it is consuming the zoe invite purse's current amount notifier, and when the pending promise for the next result is severed by upgrade, the `observeNotifier` consumer noisily reconnects.  It would be worthwhile to make this process less noisy, if possible.

File: packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/actions.sh [hunk 3]
```diff
@@ -2,4 +2,8 @@
 
 . ./upgrade-test-scripts/env_setup.sh
 
-# Core-eval contract upgrade
+# CWD is agoric-sdk
+here=./packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11
```
Comment (michaelfig) [inline packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/actions.sh]: The upgrade-test-scripts are copied to the CWD.

```suggestion
here=./upgrade-test-scripts/agoric-upgrade-11
```

File: packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/pre_test.sh [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+. ./upgrade-test-scripts/env_setup.sh
+
+echo Wait for upgrade to settle
+waitForBlock 5
+
+# CWD is agoric-sdk
+here=./packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11
```
Comment (michaelfig) [inline packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/pre_test.sh]:
```suggestion
here=./upgrade-test-scripts/agoric-upgrade-11
```

File: packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/test.sh [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+. ./upgrade-test-scripts/env_setup.sh
+
+echo Wait for actions to settle
+waitForBlock 2
+
+# CWD is agoric-sdk
+here=./packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11
```
Comment (michaelfig) [inline packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/test.sh]:
```suggestion
here=./upgrade-test-scripts/agoric-upgrade-11
```

File: packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/zoe-upgrade/vat-status.mjs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+const swingstorePath = '~/.agoric/data/agoric/swingstore.sqlite';
+
+/**
+ * SQL short-hand
+ *
+ * @param {import('better-sqlite3').Database} db
+ */
+const dbTool = db => {
+  const prepare = (strings, ...params) => {
+    const dml = strings.join(' ? ');
```
Comment (michaelfig) [inline packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/zoe-upgrade/vat-status.mjs]: Not sure why you need to insert extra whitespace beyond that which the template author used.

```suggestion
    const dml = strings.join('?');
```

File: packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/zoe-upgrade/zoe-upgrade-driver.sh (hunk 2 above)
Comment (michaelfig) [inline packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/zoe-upgrade/zoe-upgrade-driver.sh]: I think you'll need `-b block` or else your query on the next line will race and fail to display the submitted proposal.

```suggestion
  --yes -o json --from=validator --keyring-backend=test -b block
```

**Review by michaelfig — APPROVED** [review]
A few minor nits, but overall it looks good and I think it should land!

File: packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/actions.sh (hunk 3 above)
Comment (dckc) [reply packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/actions.sh]: I originally tried it that way and it failed. The fact that it passes in ci currently is pretty good evidence that it should stay the way it is, no?

Comment (dckc) [reply packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-11/actions.sh]: oh. copied. hm.

**Pull request merged**